apps_download/models/product_product.py

```diff
@@ -0,0 +1,137 @@
+# Copyright (C) 2017-Today: Odoo Community Association (OCA)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+
+import os
+import tempfile
+import shutil
+import logging
+import base64
+import time
+from odoo.exceptions import ValidationError
+from odoo import models, fields, api, _
+_logger = logging.getLogger(__name__)
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    dependent_product_ids = fields.Many2many(
+        'product.product', 'product_product_dependent_rel',
+        'src_id', 'dest_id', string='Dependent Products'
+    )
+    module_path = fields.Char(
+        related="odoo_module_version_id.repository_branch_id.local_path",
+        readonly=True)
+
+    @api.constrains('dependent_product_ids')
+    def check_dependent_recursion(self):
+        if not self._check_m2m_recursion('dependent_product_ids'):
+            raise ValidationError(
+                _('Error: You cannot create recursive dependency.')
+            )
+
+    @api.model
+    def child_dependency(self, children):
+        res = self.env['product.product']
+        for child in children:
+            if not child.dependent_product_ids:
+                continue
+            res |= child.dependent_product_ids
+            res |= self.child_dependency(child.dependent_product_ids)
+        return res
+
+    @api.multi
+    def create_dependency_list(self):
+        ret_val = {}
+        for product in self:
+            ret_val[product.id] = product.dependent_product_ids
+            if product.dependent_product_ids:
+                ret_val[product.id] |= self.child_dependency(
+                    product.dependent_product_ids)
+        return ret_val
+
+    @api.multi
+    def generate_zip_file(self):
```

Comment:
Seems this method is only usable in tests with product demo data. If it's intended to not use it for production and it must be there for some reason, I suggest to rename it to something like `demo_generate_zip_file`

Reply:
@yvaucher  This method is called from Phantom test cases and also from a Regular cron job.
only the static references in that method are for the Test case.

Reply:
@bizzappdev then don't we have an issue with the 2 products:

`apps_download.product_product_100` and `apps_download.product_product100_b`

Those 2 products should in my opinion be demo data. I don't see how the Regular cron job wouldn't fail using this method in a non demo context without the 2 previous products not installed.

I also see the method used on an action on product form
```
<button string="Get all Dependencies" type="object" name="generate_zip_file"/>
```

My main concern being, shouldn't it work without those demo products?
What will happen if I click on that "Get all Dependencies" button if I didn't installed the demo data?

If this method is only executed by PhantomJS, could you point more precisely what is executed in a real case?

What do you mean by static reference? Is it the entry in ir.model.data aka XML id ?

Reply:
@yvaucher you are right. as it is demo data it is raising an error as it is now able to find the reference in the database without demo data.

what do you suggest the best way to solve this?
1. Try and catch.
2. just remove the PhantomJS (As I also do not like to add those hardcoded references just for test cases, But in that case, we will lose the coverage)
3. any other suggestion from your side.

Note: not only from that button but also from the cron job it is calling this method and generating the Zip file for all modules/Products.

"static reference" I mean the "apps_download.product_product_100 and apps_download.product_product100_b" as this references are just for the phantomJS. for normal products it will not be used and instead it will use the module path generated fro the github interface module.

Reply:
@bizzappdev If you can't use mock and demo data to mock out your requirements in test then I am in favour of option 2.

Reply:
@gdgellatly okay sure we will do the same. cc @yvaucher  @elicoidal

Reply:
changes are committed.

Reply:
@bizzappdev See my new comment on the tests about how to mock it.